### PR TITLE
[TCL] Added TCL-Based SDC Parser

### PR DESCRIPTION
**Pull request by AlexandreSinger**
The original SDC parser was based on Flex and Bison, which were fast and worked well; however, SDC is implicitly a TCL file. Therefore, in order to be compliant with SDC (i.e. that we can support any SDC file), we need to use a real TCL interpreter.

Built a proper TCL interpreter using SWIG to define the interface. This allows LibSDCParse to parse even more SDC commands as well as other syntax it was not able to do previously (such as building procedures and defining variables).

Also added a testing infrastructure based on LLVM's Light Integration Testing (LIT) framework. This provides a very strong testing framework to ensure that the parsing is being performed correctly.

There are only two syntax coverage differences between the old SDC parser and the new SDC parser:
1. get commands must have a space between them and their arguments. Previously LibSDCParse supported something like "get_ports{...}", which is not something that many SDC parsers support.
2. Comments and the same line must have a semicolon before the '#'.

## Summary by CodeRabbit

* **New Features**
  * Optional TCL-based SDC parsing (legacy compatibility), TCL bindings and embedded TCL wrapper
  * Generated-clock support (divide/multiply) and expanded clock creation options
  * Strongly-typed object database for ports/clocks/pins/cells
  * Additional clock-group types and more consistent, flushed diagnostics/warnings

* **Chores**
  * Build/test workflow migrated to lit-based targets and test launcher; old shell harness removed

* **Tests**
  * Large expansion of automated tests and new test suites (basic, strong, legacy, doc_examples)

* **Documentation**
  * New developer README describing TCL integration and test infrastructure

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Comment by AlexandreSinger** [comment]
@coderabbitai review

File: src/tcl/sdc_commands.cpp [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+#include <cassert>
+#include <string>
+
+#include "sdcparse.hpp"
+#include "sdc_commands.h"
+#include "tcl_interpreter.h"
+
+sdcparse::Callback* sdcparse::g_callback = nullptr;
+
+void _libsdcparse_lineno_internal(int line_num) {
```
Comment (AmirhosseinPoolad) [inline src/tcl/sdc_commands.cpp]: I don't think the standard library will ever add a symbol named "_libsdcparse_lineno_internal", but it is nonetheless not best practice to start function/global variable names with an underscore. Names that start with and underscore are reserved for implementation in global scope:

>  [Each](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2014/n4296.pdf) identifier that begins with an underscore is reserved to the implementation for use as a name in the global namespace.

Same for all other function names here.

File: src/tcl/sdc_commands.cpp [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+    assert(sdcparse::g_callback != nullptr);
```
Comment (AmirhosseinPoolad) [inline src/tcl/sdc_commands.cpp]: Be careful that 'assert' will be optimized into a NOP in release builds. Not sure if it's intended or not. If you don't want to pull in dependencies from VTR, you essentially need to copy VTR's implementation of VTR_ASSERT.

File: src/tcl/sdc_commands.cpp [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+
+    // TODO: Assert that all targets are IDs
+    create_clock_cmd.targets.strings = targets;
+    create_clock_cmd.targets.type = sdcparse::StringGroupType::OBJECT;
+
+    create_clock_cmd.add = add;
+
+    assert(sdcparse::g_callback != nullptr);
+    sdcparse::g_callback->create_clock(create_clock_cmd);
+}
```
Comment (AmirhosseinPoolad) [inline src/tcl/sdc_commands.cpp:59]: I think it would be safer if this (other than line 45) was sdcparse::CreateClock's constructor.

File: src/tcl/sdc_commands.cpp [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+                                                  const std::vector<std::string>& targets) {
+    sdcparse::CreateGeneratedClock create_gen_clock_cmd;
+    create_gen_clock_cmd.name = name;
+    create_gen_clock_cmd.source = source;
+    create_gen_clock_cmd.source_string_type = sdcparse::StringGroupType::OBJECT;
+    create_gen_clock_cmd.divide_by = divide_by;
+    create_gen_clock_cmd.multiply_by = multiply_by;
+    create_gen_clock_cmd.add = add;
+    create_gen_clock_cmd.targets.strings = targets;
+    create_gen_clock_cmd.targets.type = sdcparse::StringGroupType::OBJECT;
```
Comment (AmirhosseinPoolad) [inline src/tcl/sdc_commands.cpp:75]: This also just looks like a constructor. These structs are already not C compliant (Because of the default values) so might as well use c++ to safely initialize them.

Comment (AmirhosseinPoolad) [reply src/tcl/sdc_commands.cpp:75]: My worry is that if someone decides to add a new field to the struct, it's very likely that they forget to update this function and get a weird error because of uninitialized fields. If you don't want to go all the way there with constructors and RAII, at least initialize all fields using the curly brace syntax so that the [compiler generates a warning if you leave a field uninitialized](https://godbolt.org/z/s8hj5YfYd).

Update: The compiler only warns with curly braces syntax if the fields don't have a default value!

File: src/tcl/sdc_commands.cpp [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+    sdcparse::g_callback->set_clock_groups(set_clock_groups_cmd);
+}
+
+void _libsdcparse_set_false_path_internal(const std::vector<std::string>& from_list,
+                                          const std::vector<std::string>& to_list) {
+    sdcparse::SetFalsePath set_false_path_cmd;
+    set_false_path_cmd.from.strings = from_list;
+    set_false_path_cmd.from.type = sdcparse::StringGroupType::OBJECT;
+    set_false_path_cmd.to.strings = to_list;
+    set_false_path_cmd.to.type = sdcparse::StringGroupType::OBJECT;
```
Comment (AmirhosseinPoolad) [inline src/tcl/sdc_commands.cpp:116]: Same initialization unsafety issues as before.

Comment (AmirhosseinPoolad) [reply src/tcl/sdc_commands.cpp:116]: This pattern exists throughout this file. I won't spam you with all of them.

File: src/tcl/sdc_commands.cpp [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+    set_timing_derate_cmd.cell_targets.strings = targets;
+    set_timing_derate_cmd.cell_targets.type = sdcparse::StringGroupType::OBJECT;
+
+    assert(sdcparse::g_callback != nullptr);
+    sdcparse::g_callback->set_timing_derate(set_timing_derate_cmd);
+}
+
+std::vector<std::string> _libsdcparse_all_ports_internal() {
+    assert(sdcparse::g_callback != nullptr);
+    return sdcparse::g_callback->obj_database.get_port_objects();
```
Comment (AmirhosseinPoolad) [inline src/tcl/sdc_commands.cpp:280]: This function (and get_port_objects), returns a copy of the vector. Why not have them both return a const reference? The API users could make a copy if they wanted to.

File: src/tcl/sdc_commands.cpp [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+
+std::vector<std::string> _libsdcparse_all_pins_internal() {
+    assert(sdcparse::g_callback != nullptr);
+    return sdcparse::g_callback->obj_database.get_pin_objects();
+}
+
+std::vector<std::string> _libsdcparse_all_cells_internal() {
+    assert(sdcparse::g_callback != nullptr);
+    return sdcparse::g_callback->obj_database.get_cell_objects();
+}
```
Comment (AmirhosseinPoolad) [inline src/tcl/sdc_commands.cpp]: ditto for these two functions.

File: src/tcl/sdc_commands.h [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+/**
+ * @brief Raise a warning message from the TCL side.
+ */
+void _libsdcparse_raise_warning(const std::string& msg);
+
+// The following are internal wrapper calls to the underlying callback data
+// structures. The line up basically one-to-one with the SDC commands that are
+// supported.
+
+void _libsdcparse_create_clock_internal(double period,
```
Comment (AmirhosseinPoolad) [inline src/tcl/sdc_commands.h]: These functions are internal but you should still document them. Won't spam you for every single function in this file.

File: src/tcl/tcl_interpreter.h [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+#include <string>
+
+#include "sdc_wrapper.h"
+#include "sdcparse.hpp"
+
+extern "C" {
+    /**
+     * @brief Initialization command for the TCL interface generated by SWIG.
+     */
+    extern int Sdc_commands_Init(Tcl_Interp* interp);
```
Comment (AmirhosseinPoolad) [inline src/tcl/tcl_interpreter.h:23]: Is it a SWIG thing that the S in Sdc is capitalized?

File: src/main.cpp [hunk 10]
```diff
@@ -20,21 +22,109 @@ class PrintCallback : public Callback {
     void lineno(int line_num) override { lineno_ = line_num; }
 
     void create_clock(const CreateClock& cmd) override {
+        fflush(stdout);
```
Comment (AmirhosseinPoolad) [inline src/main.cpp]: Why do you need to flush stdout? It messes up the internal OS buffering and is not considered best practice.

File: src/main.cpp [hunk 11]
```diff
@@ -1,3 +1,5 @@
+#include <iostream>
```
Comment (AmirhosseinPoolad) [inline src/main.cpp]: iostream is only used in line 297 (Which also uses std::endl which is not considered best practice anymore and everything else uses stdio.h and printf. I think you might as well change that to printf and get rid of this include.

File: src/main.cpp [hunk 12]
```diff
@@ -20,21 +22,109 @@ class PrintCallback : public Callback {
...
+        } else if (cmd.targets.strings.size() == 1) {
+            if (cmd.targets.type == StringGroupType::OBJECT) {
+                clock_name = obj_database.get_object_name(ObjectId(cmd.targets.strings[0]));
+            } else {
+                clock_name = cmd.targets.strings[0];
+            }
+            obj_database.create_clock_object(clock_name);
+        } else {
+            // If no name is given, create a clock for each target.
+            // TODO: Verify that this is correct.
```
Comment (AmirhosseinPoolad) [inline src/main.cpp]: Reminder to verify, if you haven't.

File: src/main.cpp [hunk 13]
```diff
@@ -186,16 +308,27 @@ class PrintCallback : public Callback {
...
         fprintf(stderr, "Reads in an SDC file into internal data structures\n");
-        fprintf(stderr, "and then prints it out\n");
-        exit(1);
+        fprintf(stderr, "and then prints it out.\n");
+        fprintf(stderr, "Use -legacy to use the legacy (non-tcl) parser.\n");
+        return 1;
+    }
+
+    bool use_tcl_interp = true;
+    for (int i = 2; i < argc; i++) {
```
Comment (AmirhosseinPoolad) [inline src/main.cpp:305]: This loop is a bit weird since it's actually only an if statement. I think you should just check if argc > 2 and parse argv[2] directly instead of this.

File: src/sdc_timing_object_database.h [hunk 14]
```diff
@@ -0,0 +1,277 @@
...
+     * @brief Create a cell object.
+     *
+     *  @param cell_name
+     *      The name of the port to create. This does not need to be unique.
+     *
+     *  @return The ID of the created object.
+     */
+    CellObjectId create_cell_object(const std::string& cell_name) {
+        CellObjectId cell_object_id = CellObjectId("__vtr_obj_cell_" + std::to_string(cell_objects.size()));
+        assert(object_name.count(cell_object_id) == 0);
```
Comment (AmirhosseinPoolad) [inline src/sdc_timing_object_database.h:76]: Is this error also checked in the TCL side?

Comment (AmirhosseinPoolad) [reply src/sdc_timing_object_database.h:76]: Same with all other asserts in this file.

File: src/sdc_timing_object_database.h [hunk 15]
```diff
@@ -0,0 +1,277 @@
+#pragma once
```
Comment (AmirhosseinPoolad) [inline src/sdc_timing_object_database.h:1]: Throughout this whole file all function definitions are in the .h file. Is this intended?

File: test/__pycache__/lit.cpython-312.pyc
Comment (AmirhosseinPoolad) [inline test/__pycache__/lit.cpython-312.pyc:1]: Delete this

File: .gitignore [hunk 16]
```diff
@@ -7,3 +7,9 @@ src/sdc_lexer.gen.h
 src/sdc_parser.gen.c
 src/sdc_parser.gen.h
 src/stack.hh
+
+.cache/
+
+.venv/
+
+__pycache__/
```
Comment (AmirhosseinPoolad) [inline .gitignore:15]: Should probably also add test/__pycache__ here

**Comment by AmirhosseinPoolad** [comment]
Thanks for the PR Alex! Did not review any of the TCL/SDC files as I don't know TCL but the C++ side looked good!

File: src/tcl/sdc_commands.cpp (hunk 4 above)
Comment (AlexandreSinger) [reply src/tcl/sdc_commands.cpp:75]: There is another issue with that approach. The issue is that it relies on the order that the members are specified in the struct. This means that if someone were to reorder the struct, it may not generate a warning or an error if the types happen to be the same! In my opinion, that is a worse thing to have.

File: src/main.cpp (hunk 10 above)
Comment (AlexandreSinger) [reply src/main.cpp]: AH! This was the source of some of the WORST bugs I was dealing with in the unit tests!

The issue comes from how file buffers work. Since TCL and this main.cpp file use different buffers, the print messages actually get reordered. This was causing issues in my unit tests where I relied on messages being generated on both the TCL and the main.cpp side.

This file is only used for testing, so I am ok with flushing the stdout all these times.

Comment (AmirhosseinPoolad) [reply src/main.cpp]: Understandable. May I propose wrapping the flush->print->flush cycle in a macro?

```
#define flushing_printf(format, ...) do { \
    fflush(stdout);                   \
    printf(format, ##__VA_ARGS__);    \
    fflush(stdout);                   \
} while (0)
```

Comment (AlexandreSinger) [reply src/main.cpp]: This was a good idea. I updated it with a similar approach.

File: src/main.cpp (hunk 13 above)
Comment (AlexandreSinger) [reply src/main.cpp:305]: I wanted to future-proof this function to allow us to add more flags if needed.

File: src/tcl/sdc_commands.cpp (hunk 1 above)
Comment (AlexandreSinger) [reply src/tcl/sdc_commands.cpp]: I agree with this point. I originally wanted to avoid collisions on the SDC side; however the libsdcparse_ prefix should be unique enough. Good point.

File: src/tcl/sdc_commands.cpp (hunk 2 above)
Comment (AlexandreSinger) [reply src/tcl/sdc_commands.cpp]: These asserts should be correct by construction of this tool. They should never be tripped; so I only want them on for debugging.

File: src/tcl/sdc_commands.cpp (hunk 6 above)
Comment (AlexandreSinger) [reply src/tcl/sdc_commands.cpp:280]: The lifetime of that vector would be unknown since the object is passed into the TCL side. I am just nervous of passing that information. For now, the performance is not an issue.

File: src/sdc_timing_object_database.h (hunk 14 above)
Comment (AlexandreSinger) [reply src/sdc_timing_object_database.h:76]: These are similar to the g_callback asserts; they should be correct by construction. I only want these checks to occur during debug checks.

File: src/sdc_timing_object_database.h (hunk 15 above)
Comment (AlexandreSinger) [reply src/sdc_timing_object_database.h:1]: This is intended. All of these methods are basically getters and setters. To me, it makes the header file more clear and self-documenting since all of these methods are less than 5 lines each. I added the 'inline' keyword to make this more explicit.

File: src/tcl/sdc_commands.h (hunk 8 above)
Comment (AlexandreSinger) [reply src/tcl/sdc_commands.h]: The comment above is meant to be all encompassing. These functions should only ever be called by TCL. They are not intended for use on the CPP side.

**Comment by AlexandreSinger** [comment]
@coderabbitai review

File: src/tcl/script_converter.py [hunk 17]
```diff
@@ -45,12 +45,11 @@
 
                 # Some special characters need to be properly terminated to make
                 # a valid C-string.
+                line = line.replace('\\', '\\\\')
                 line = line.replace('\"', '\\"')
```
Comment (AmirhosseinPoolad) [inline src/tcl/script_converter.py:49]: Does this not result in too many slashes?

Image

Comment (AlexandreSinger) [reply src/tcl/script_converter.py:49]: Resolved in person.

**Comment by AlexandreSinger** [comment]
Added @petergrossmann21 to help review the TCL changes.

@vaughnbetz Please let me know if you would like to review the docs. I have added some developer documentation to explain some of the design decisions I have made for this.

File: test/basic/set_false_path.sdc
Comment (petergrossmann21) [inline test/basic/set_false_path.sdc:1]: Add a section to this test where set_false_path has startpoint/endpoints that are not clocks; I know VPR doesn't support this yet but it's a key addition that we're looking for.

File: test/basic/set_input_delay.sdc
Comment (petergrossmann21) [inline test/basic/set_input_delay.sdc:1]: Add one check here where -clock is omitted.

File: test/basic/set_max_delay.sdc
Comment (petergrossmann21) [inline test/basic/set_max_delay.sdc:1]: Add tests here where only one of -from and -to are specified

File: test/basic/set_min_delay.sdc
Comment (petergrossmann21) [inline test/basic/set_min_delay.sdc:1]: Add tests here where only one of -from and -to are specified

File: test/basic/set_multicycle_path.sdc
Comment (petergrossmann21) [inline test/basic/set_multicycle_path.sdc:1]: Add tests here where only one of -from and -to are specified

**Review by petergrossmann21 — COMMENTED** [review]
Are legacy and old_tests slated for deletion?  What do they provide if anything that your basic and strong tests do not?

Other than that, it looks like just filling in a few more corner cases can round out your test suite; see inline comments.

File: docs/dev/README.md [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+Synopsys Design Constraints (SDCs) are a set of TCL procedures that a user can call to constrain parts of their ASIC designs. SDC has been adopted by FPGA vendors to allow users to constrain aspects of their design (such as the clock frequencies). There are many SDC procedures that are not necessary for FPGAs, and so are often not supported. For FPGA designs, we are most interested in three sets of constraints:
+1. Timing Constraints
+2. Timing Exceptions
+3. Object Queries
+
+See the following documentation for more information: [FPGA Timing Suite Syntax Docs](https://github.com/siliconcompiler/fpga-timing-coverage-suite/tree/master/docs/syntax).
+
+### The TCL Interpreter in C++
+
+Due to SDC being necessarily a TCL file, SDCs are actually interpretted. This means that an SDC file is not executed line-by-line, but is executed similar to how Python is executed. As such, it is necessary for a fully SDC compliant parser to be a TCL interpreter.
```
Comment (vaughnbetz) [inline docs/dev/README.md]: similar -> similarly

File: docs/dev/README.md [hunk 19]
```diff
@@ -0,0 +1,133 @@
...
+
+Fortunately, it is somewhat trivial to construct a TCL interpreter since it is provided as a very well documented library. The [TCL/Tk](https://www.tcl-lang.org/) toolkits provide C APIs for creating and invoking a TCL interpreter. See the following [wiki page](https://wiki.tcl-lang.org/page/Adding+Tcl%2FTk+to+a+C+application) for more information.
+
+The entire TCL interpreter for the SDC parser in this repository can be found in:
```
Comment (vaughnbetz) [inline docs/dev/README.md]: entire interpreter seems a bit confusing.  I think you want to say the code to instantiate an interpreter (using the library above) is in ...

File: docs/dev/README.md [hunk 20]
```diff
@@ -0,0 +1,133 @@
...
+    tcl/
+        sdc_commands.cpp
+        sdc_commands.h
+        sdc_commands.i
+```
+The `sdc_commands.i` file is a SWIG interface file. It defines how the C++ functions in sdc_commands.h will be turned into TCL procedures which can be called by our wrapper file. Each of these commands are "internal" commands which we do not expect a user to use directly on the TCL side.
+
+The TCL interface procedures are generated during build time. This process is described within the top level `CMakeLists.txt` file.
+
+The functions defined in `sdc_commands.cpp` is what actually takes the information from the parser and passes it into the callback. Due to how SWIG is set up, the most convenient way to do this is by creating a global callback object pointer which is initialized by the interpreter and used within these functions.
```
Comment (vaughnbetz) [inline docs/dev/README.md]: is what actually takes ->are what actually take

File: docs/dev/README.md [hunk 21]
```diff
@@ -0,0 +1,133 @@
...
+```txt
+src/
+    sdc_timing_object_database.h
+    sdc_timing_object.h
+```
+This database is accessible by the tool (i.e. VPR), allowing it create different types of objects that it needs. For example, before parsing the SDC file, it is expected that VPR would create port objects for all ports in the netlist. The parser will then use that information to manage what to do when an SDC file contains the `get_ports` procedure.
+
+All timing objects have unique IDs which act like pointers in TCL. To make the interface as easy as possible, on the TCL side, these objects appear as strings with a unique format: `__vtr_obj_<type>_<id>`. This is used to encode information about the object to make debugging easier. This is different from how SDC parser normally store these objects. Normally, these objects are stored as pointers to TCL objects. This is because these objects are accessed by other algorithms and are timing critical. Since the goal of this code is to just act like a parser, it is not as timing critical. So we would prefer the simplicity and debuggability of using strings instead.
```
Comment (vaughnbetz) [inline docs/dev/README.md]: how SDC parser -> how other SDC parsers normally store

File: docs/dev/README.md [hunk 22]
```diff
@@ -0,0 +1,133 @@
...
+
+### Overview of TCL-Based SDC Parser
+
+The following figure provides an overview of how the different components of the TCL interpreter interact with one-another:
+
+![TCL-Based SDC Parser Overview](../figures/LibSDCParse-TCL-overview.png)
+
+The TCL interpreter begins by parsing the SWIG interface commands (defined in C++ and then generated by SWIG). It then loads the SDC wrapper file (which is turned into a string by the script converter). It then parses the SDC file (which is provided directly by the user).
```
Comment (vaughnbetz) [inline docs/dev/README.md]: parsing -> registering

Comment (vaughnbetz) [reply docs/dev/README.md]: loads the SDC wrapper -> parses and loads the SDC wrapper

File: docs/dev/README.md [hunk 23]
```diff
@@ -0,0 +1,133 @@
...
+
+In this example, when `set_multicycle_path` is parsed, the procedure in `sdc_wrapper.tcl` is called. This parses the arguments of the SDC commands, check that it is valid, and then passes the arguments to the internal SWIG interface function. This interface function then calls the `set_multicycle_path` callback which will run code in the tool (in this case, VPR).
```
Comment (vaughnbetz) [inline docs/dev/README.md]: check that it is -> checks that they are 
(or say it checks the overall command is valid)

File: docs/dev/README.md [hunk 24]
```diff
@@ -0,0 +1,133 @@
...
+make check-libsdcparse
+```
+This will run the entire test suite (using as many cores as available).
+
+There is another Python script found in the test folder: `test/not.py`. This is a helper script based on the `not` utility executable in LLVM. This Python script negates the return of an executable such that when the executable fails (returns 1), this script returns 0, and vice-versa. This allows us to create test cases which are "expected to fail", such as syntax errors.
+
+There is also a suppression file, `test/tcl.supp`, which is used to suppress known false-positives for Valgrind. Valgrind can be run using a custom executable defined in `test/CMakeLists.txt`:
+```sh
+make check-libsdcparse-vg
+```
```
Comment (vaughnbetz) [inline docs/dev/README.md:126]: Testing: you should check that vpr is valgrind clean after this is merged and integrated into vpr.  Possibly you'll have to do something to get this added to vpr's suppression file list.

File: docs/dev/README.md [hunk 25]
```diff
@@ -0,0 +1,133 @@
...
+
+All other directories in the `test` folder will be traversed in search for test cases. All files which end in `.sdc` will be executed as a testcase.
+
+At the top of each testcase, you will find a `RUN` command. This command tells LIT how to execute that script. See LLVM-LIT's documentation for more information on how this works.
+
+In each testcase, you will find `CHECK` commands. These commands are used by FileCheck which ensures that the given line can be found in stdout. See the [FileCheck Docs](https://llvm.org/docs/CommandGuide/FileCheck.html) for more information; but in a nutshell, we are able to use pattern-matching to ensure that the correct output is generated by our testing scripts.
+
+To run our testcases, we use a test executable called `sdcparse_test`. This executable is defined in `src/main.cpp`. This executable pretends like it is a downstream tool and whenever it sees a command, it prints that command and all of its arguments to standard out. In our testcases, we just ensure that the command that we put in is the command that is printed out.
```
Comment (vaughnbetz) [inline docs/dev/README.md]: I'd add a quick description of how to add a test.  Step 1 ...
Don't reproduce a bunch of existing docs, but give an overview and links if appropriate.

**Review by vaughnbetz — CHANGES_REQUESTED** [review]
The documentation looks very good; just a few minor nits.
Think about making a new tab in the vtr documentation for this library / repo, once this is merged (seems like it would be good to have a tab to make the Doxygen and overview more accessible).

**Comment by AlexandreSinger** [comment]
> Are legacy and old_tests slated for deletion? What do they provide if anything that your basic and strong tests do not?
> 
> Other than that, it looks like just filling in a few more corner cases can round out your test suite; see inline comments.

The legacy tests use the original Flex+Bison parser and are slated for deletion. The old_tests are the old legacy tests which I wanted to keep around to ensure that the old syntax expected from users of this library is still supported. They should not provide anything that the basic and strong do not; but they just ensure that we are not regressing the features of LibSDCParse.

Thanks for the comments, I agree with them and will add the tests.

**Comment by AlexandreSinger** [comment]
@coderabbitai review
